In [1]:
import macpath

from fastapi import FastAPI

app = FastAPI()


@app.get('/')
async def hello_world():
    return {"hello": "world!"}

In [5]:
import uvicorn

uvicorn.run(app, host="0.0.0.0", port=8000)

C:\Users\Azat\Developer\FastAPI\fastapi-data-science\venv\lib\site-packages\traitlets\traitlets.py:1045: RuntimeWarning: coroutine 'Server.serve' was never awaited
  for key in dir(cls):


RuntimeError: asyncio.run() cannot be called from a running event loop

In [6]:
!pip3 install nest_asyncio

In [5]:
import nest_asyncio

nest_asyncio.apply()

In [10]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:9124 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:9124 - "GET /docs/ HTTP/1.1" 307 Temporary Redirect
INFO:     127.0.0.1:9124 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:9124 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:9125 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:9329 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:9381 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:9381 - "GET /openapi.json HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218134800.png)

![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218134915.png)

## Handling request parameters

In [11]:
from fastapi import FastAPI

app = FastAPI()


@app.get('/users/{user_id}')
async def get_user(user_id: int):
    return {"id": user_id}

In [12]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:9517 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:9517 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:9518 - "GET /users/5 HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


In [13]:
from fastapi import FastAPI

app = FastAPI()


@app.get('/users/{user_type}/{user_id}')
async def get_user(user_type: str, user_id: int):
    return {"user_type": user_type, "id": user_id}

In [14]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:9578 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:9578 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:9579 - "GET /users/student/8 HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


### Limiting allowed values

Once again, we'll lean on type
hinting. Python has a very useful class for this: Enum. An enumeration is a way to list
all the valid values for a specific kind of data. Let's define an Enum class that will list the
different types of users:

In [16]:
from enum import Enum
from fastapi import FastAPI


class UserType(str, Enum):
    STANDARD = "standard"
    ADMIN = "admin"

In [17]:
from fastapi import FastAPI

app = FastAPI()


@app.get('/users/{user_type}/{user_id}')
async def get_user(user_type: UserType, user_id: int):
    return {"user_type": user_type, "id": user_id}

In [18]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:9665 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:9665 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:9666 - "GET /users/standard/5 HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


### Advanced validation

In [19]:
from fastapi import FastAPI, Path

app = FastAPI()


@app.get('/users/{user_id}')
async def get_user(user_id: int = Path(..., ge=1)):
    return {'user_id': user_id}

In [20]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:10350 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:10350 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:10351 - "GET /users/1 HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


There are several things to pay attention to here: the result of Path is used as a default
value for the id argument in the path operation function.

• gt: Greater than
• ge: Greater than or equal to
• lt: Less than
• le: Less than or equal to

There are also validation options for string values, which are based on the length and the
use of a regular expression.

In [21]:
from fastapi import FastAPI, Path

app = FastAPI()


@app.get('/license-plates/{_license}')
async def get_license_plate(_license: str = Path(..., min_length=9, max_length=9)):
    return {'license': _license}

In [22]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:10451 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:10451 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:10452 - "GET /license-plates/ABCDEFGHF HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


In [23]:
# using regx
from fastapi import FastAPI, Path

app = FastAPI()


@app.get('/license-plates/{_license}')
async def get_license_plate(_license: str = Path(..., regex=r"^\w{2}-\d{3}-w{2}$")):
    return {'license': _license}

In [24]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


Thanks to this regular expression, we only accept strings that exactly match the license
plate format. Notice that the regular expression is prefixed with r. Just like f-strings,
this is a Python syntax that is used to indicate that the following string should be
considered as a regular expression.

### Query parameters

Query parameters are a common way to add some dynamic parameters to a URL. You
find them at the end of the URL in the following form: ?param1=foo&param2=bar.
In a REST API, they are commonly used on read endpoints to apply pagination, a filter, a
sorting order, or selecting fields

In [25]:
@app.get('/users')
async def get_user(page: int = 1, size: int = 10):
    return {"page": page, "size": size}

In [26]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:10755 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:10755 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:10756 - "GET /users?page=1&size=10 HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


In [27]:
class UserFormat(str, Enum):
    SHORT = "short"
    FULL = "full"


app = FastAPI()


@app.get("/users")
async def get_users(_format: UserFormat):
    return {"format": _format}

In [28]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:10885 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:10885 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:10885 - "GET /users?_format=full HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


We also have access to more advanced validations through the Query function. It works
in the same way that we demonstrated in the Path parameters section:

In [29]:
from fastapi import Query

app = FastAPI()


@app.get('/users')
async def get_user(page: int = Query(1, gt=0), size: int = Query(10, le=1000)):
    return {'page': page, 'size': size}

In [30]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:14573 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:14573 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:14574 - "GET /users?page=1&size=222 HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


the default parameter value is the first argument of the Query function.

### The request body

The body is the part of the HTTP request that contains raw data, representing documents,
files, or form submissions. In a REST API, it's usually encoded in JSON and used to create
structured objects in a database.

For the simplest cases, retrieving data from the body works exactly like query parameters.
The only difference is that you always have to use the
function; otherwise, FastAPI
Body
will look for it inside the query parameters by default

In [34]:
from fastapi import Body

app = FastAPI()


@app.post("/users")
async def create_user(name: str = Body(...), age: int = Body(...)):
    return {
        "name": name,
        'age': age
    }

In [35]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:14726 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:14726 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:14727 - "POST /users HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


You also have access to more advanced validation through the Body function. It works in
the same way as we demonstrated in the Path parameters section.

This is why FastAPI uses pydantic models for data validation. Pydantic is a Python library
for data validation and is based on classes and type hints. In fact, the Path, Query, and
Body functions that we've learned about so far use pydantic under the hood!

In [1]:
from fastapi import FastAPI

from pydantic import BaseModel


class User(BaseModel):
    name: str
    age: int


app = FastAPI()


@app.post("/users")
async def create_user(user: User):
    return user

In [37]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7508]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:14846 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:14846 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:14847 - "POST /users HTTP/1.1" 422 Unprocessable Entity


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7508]


### Multiple Objects

In [39]:
class Company(BaseModel):
    name: str


@app.post("/users")
async def create_user(user: User, company: Company):
    return {
        "user": user,
        'company': company
    }

 You can even add singular body values with the Body function, just as we saw at the
beginning of this section. This is useful if you wish to have a single property that's not
part of any model:

In [40]:
@app.post("users/")
async def create_user(user: User, priority: int = Body(..., ge=1, le=3)):
    return {
        "user": user,
        "priority": priority
    }

The priority property is an integer between 1 and 3, which is expected beside the
user object:

echo '{"user": {"name": "John", "age": 30}, "priority": 1}' |
http POST http://localhost:8000/users

### Format data and file uploads

Even if REST APIs work most of the time with JSON, sometimes, you might have
to handle form-encoded data or file uploads, which have been encoded either as
application/x-www-form-urlencoded or multipart/form-data.

In [2]:
from fastapi import Form
import uvicorn

# form data

app = FastAPI()


@app.post("/users")
async def create_user(name: str = Form(...), age: int = Form(...)):
    return {
        'name': name,
        'age': age
    }

In [3]:
!pip install python-multipart

In [6]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7596]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:1437 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:1437 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:1438 - "POST /users HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7596]


Pay attention to how the
header and the body data representation have
Content-Type
changed in the request. You can also see that the response is still provided in JSON. Unless
specified otherwise, FastAPI will always output a JSON response by default, no matter the
form of the input data.

### File uploads

Uploading files is a common requirement for web applications, whether this is images or
documents. FastAPI provides a parameter function, File, that enables this.

you can directly retrieve a file as a bytes object:

In [7]:
from fastapi import FastAPI, File

app = FastAPI()


@app.post("/files")
async def upload_file(file: bytes = File(...)):
    return {
        "file_size": len(file)
    }

In [8]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7596]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:1531 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:1531 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:1532 - "POST /files HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7596]


Once again, you can see that the approach is still the same: we define an argument for the
path operation function, file, we add a type of hint, bytes, and then we use the
File
function as a default value for this argument. By doing this, FastAPI understands that it
will have to retrieve raw data in a part of the body named
and return it as bytes.
file

One drawback to this approach is that the uploaded file is entirely stored in memory. So,
while it'll work for small files, it is likely that you'll run into issues for larger files. Besides,
manipulating a bytes object is not always convenient for file handling.

### UploadFile class (right way to handle file)

This class will store the data
in memory up to a certain threshold and, after this, will automatically store it on disk in
a temporary location. This allows you to accept much larger files without running out of
memory.

In [9]:
from fastapi import FastAPI, File, UploadFile

app = FastAPI()


@app.post("/files")
async def upload_file(file: UploadFile = File(...)):
    return {
        "file_name": file.filename,
        "content_type": file.content_type
    }

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

In [10]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7596]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:1693 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:1693 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:1694 - "POST /files HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7596]


Notice that, here, we return the filename and content_type properties. The content
type is especially useful for checking the type of the uploaded file and possibly rejecting it if
it's not one of the types that you expect.

You can even accept multiple files by type hinting the argument as a list of UploadFile:

### Multiple Files

In [11]:
from typing import List
from fastapi import FastAPI, File, UploadFile

app = FastAPI()


@app.post("/files")
async def upload_multiple_file(files: List[UploadFile] = File(...)):
    return [
        {
            "file_name": file.filename,
            "content_type": file.content_type
        } for file in files
    ]

In [12]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [7596]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:1787 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:1787 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:1788 - "POST /files HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [7596]


## Headers and Cookie

Besides the URL and the body, another major part of the HTTP request are the headers.
They contain all sorts of metadata that can be useful when handling requests. A common
usage is to use them for authentication, for example, via the famous cookies.

In [2]:
from fastapi import FastAPI, Header

app = FastAPI()

app.get("/")


async def get_header(hello: str = Header(...)):
    return {
        "hello": hello
    }

In [4]:
import nest_asyncio
import uvicorn

nest_asyncio.apply()

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:5478 - "GET / HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:5485 - "GET / HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:5485 - "GET / HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:5502 - "GET / HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:5509 - "GET / HTTP/1.1" 404 Not Found


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


In [7]:
from fastapi import FastAPI, Header

app = FastAPI()


@app.get("/")
async def get_header(user_agent: str = Header(...)):
    return {
        "user_agent": user_agent
    }

In [8]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:5589 - "GET / HTTP/1.1" 404 Not Found


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


In [9]:
from fastapi import FastAPI, Header

app = FastAPI()


@app.get("/")
async def get_header(user_agent: str = Header(...)):
    return {"user_agent": user_agent}

In [10]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:5618 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:5627 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:5627 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:5640 - "GET / HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


One very special case of header is cookies. You could retrieve them by parsing the
Cookie header yourself, but that would be a bit tedious. FastAPI provides another
parameter function that automatically does it for you.

In [14]:
from typing import Optional
from fastapi import FastAPI, Cookie

app = FastAPI()


@app.get("/")
async def get_header(hello: Optional[str] = Cookie(None)):
    '''
    Notice that we type hinted the argument as Optional, and we set a default value of to the function. This way, even if the cookie is not set in the request,None Cookie
    FastAPI will proceed and not generate a 422 status error response.
    '''
    return {"user_agent": hello}

In [12]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:5723 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:5729 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:5762 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:5778 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:5800 - "GET / HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218184131.png)

返回结果:
> GET / HTTP/1.1
> Host: localhost:8000
> User-Agent: insomnia/2021.7.2
> Cookie: hello=bar
> Hello: World
> Accept: */*

```
{
	"user_agent": "bar"
}
```

### The request Object
Sometimes, you might find that you need to access a raw request object with all of the data
associated with it.

In [15]:
from fastapi import FastAPI, Request

app = FastAPI()


@app.get("/")
async def get_request_object(request: Request):
    return {
        "path": request.url.path
    }

In [16]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:5970 - "GET / HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


{
	"path": "/"
}

## Customizing the response

- status code
- raising validation errors
- setting cookies

customize the response declaratively by using path operation parameters.

### path operation parameters

in order to
create a new endpoint, you had to put a decorator on top of the path operation function.
This decorator accepts a lot of options, including ones to customize the response.

In [17]:
## status code


from fastapi import FastAPI, status
from pydantic import BaseModel


class Post(BaseModel):
    title: str


app = FastAPI()


@app.post("/posts", status_code=status.HTTP_201_CREATED)
async def create_post(post: Post):
    return post

In [18]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:6133 - "GET / HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:6139 - "GET /posts/ HTTP/1.1" 307 Temporary Redirect
INFO:     127.0.0.1:6139 - "GET /posts HTTP/1.1" 405 Method Not Allowed
INFO:     127.0.0.1:6144 - "POST /posts/ HTTP/1.1" 307 Temporary Redirect
INFO:     127.0.0.1:6144 - "POST /posts HTTP/1.1" 422 Unprocessable Entity
INFO:     127.0.0.1:6160 - "POST /posts/ HTTP/1.1" 307 Temporary Redirect
INFO:     127.0.0.1:6160 - "POST /posts HTTP/1.1" 201 Created


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218185053.png)

{
	"title": "post title"
}

In [19]:
## return 204 after delete content

# Dummy database

posts = {
    1: Post(title="Hello", nb_views=100),
}


@app.delete("/posts/{id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_post(id: int):
    posts.pop(id, None)
    return None

In [20]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:6288 - "POST /posts/ HTTP/1.1" 307 Temporary Redirect
INFO:     127.0.0.1:6288 - "POST /posts HTTP/1.1" 201 Created
INFO:     127.0.0.1:6288 - "DELETE /posts/ HTTP/1.1" 307 Temporary Redirect
INFO:     127.0.0.1:6288 - "DELETE /posts HTTP/1.1" 405 Method Not Allowed
INFO:     127.0.0.1:6288 - "DELETE /posts/1 HTTP/1.1" 204 No Content


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218185431.png)

Notice that you can very well return None in your path operation function. FastAPI will
take care of it and return a response with an empty body.

### The response model

With FastAPI, the main use case is to directly return a pydantic model that automatically
gets turned into properly formatted JSON.

In [25]:
from fastapi import FastAPI
from pydantic import BaseModel


class Post(BaseModel):
    title: str
    nb_views: int


app = FastAPI()

# Dummy database

posts = {
    1: Post(title="Hello", nb_views=100),
}


@app.get("/posts/{post_id}")
async def get_post(post_id: int):
    return posts[post_id]

In [26]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:6540 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:6540 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:6540 - "GET /posts/1 HTTP/1.1" 200 OK
INFO:     127.0.0.1:6579 - "GET /posts/1 HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


```json
{
	"title": "Hello",
	"nb_views": 100
}
```

In [27]:
# do not let the end user see the post view counts

from fastapi import FastAPI
from pydantic import BaseModel


class Post(BaseModel):
    title: str
    nb_views: int


class PublicPost(BaseModel):
    title: str


app = FastAPI()

# Dummy database

posts = {
    1: Post(title="Hello", nb_views=100),
}


@app.get("/posts/{post_id}", response_model=PublicPost)
async def get_post(post_id: int):
    return posts[post_id]

In [28]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:6640 - "GET /posts/1 HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


```json
{
	"title": "Hello"
}
```

### The response parameter

Sometimes, it might be useful to return some custom headers or set cookies. This can be
done dynamically using FastAPI directly within the path operation logic.

In [30]:
### Setting headers

In [31]:
from fastapi import FastAPI, Response

app = FastAPI()


@app.get("/")
async def custom_header(response: Response):
    response.headers["Custom-Header"] = "Custom-Header-Value"
    return {
        "hello": "world!"
    }

In [32]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:6782 - "GET /posts/1 HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:6786 - "GET / HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


date: Sat, 18 Dec 2021 13:05:54 GMT
server: uvicorn
content-length: 18
content-type: application/json
custom-header: Custom-Header-Value

Also, notice that you don't have to return the Response object. You can still return
JSON-encodable data and FastAPI will take care of forming a proper response, including
the headers you've set.

In [33]:
# setting Cookies

In [34]:
from fastapi import FastAPI, Response

app = FastAPI()


@app.get("/")
async def custom_cookie(response: Response):
    response.set_cookie("cookie-name", "cookie-value",
                        max_age=86400)  # it'll be valid for 86400 seconds = 86400/60 = 1440 minutes = 1440/60 = 24 hours
    return {
        "hello": "world!"
    }

In [35]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:6973 - "GET / HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


date: Sat, 18 Dec 2021 13:10:20 GMT
server: uvicorn
content-length: 18
content-type: application/json
set-cookie: cookie-name=cookie-value; Max-Age=86400; Path=/; SameSite=lax

the browser will remove the cookie after 24 hours

### setting the status code dynamically

A good approach would be to return a
status when the object
200 OK
already exists or a
status when the object has to be created.
201 Created

In [40]:
from fastapi import FastAPI
from pydantic import BaseModel


class Post(BaseModel):
    title: str
    nb_views: int


class PublicPost(BaseModel):
    title: str


app = FastAPI()

# Dummy database

posts = {
    1: Post(title="Hello", nb_views=100),
}


# update an object in the database if exists or create it if not exists.
@app.put("/posts/{post_id}")
async def update_or_create_post(post_id: int, post: PublicPost, response: Response):
    if post_id not in posts:
        response.status_code = status.HTTP_201_CREATED
    posts[post_id] = post
    return posts[post_id]

In [41]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:7269 - "PUT /posts/1 HTTP/1.1" 200 OK
INFO:     127.0.0.1:7321 - "PUT /posts/2 HTTP/1.1" 201 Created


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218191858.png)

The post with an ID of 1 already exists, so we get a 200 status. Now, let's try with a
non-existing ID:

![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218191953.png)

## Raising HTTP errors

In a REST API, there are two very
important things that you can use to return an informative message: the status code and
the payload.

To raise an HTTP error in FastAPI, you'll have to raise a Python exception,
HTTPException. This exception class will allow us to set a status code and an error
message.

In [42]:
from fastapi import FastAPI, Body, HTTPException

app = FastAPI()


@app.post("/password")
async def check_password(password: str = Body(...), password_confirm: str = Body(...)):
    if password != password_confirm:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            detail="Passwords don't match.",
        )
    return {
        "message": "Passwords match"
    }

In [43]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:13528 - "PUT /posts/2 HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:13535 - "PUT /password HTTP/1.1" 405 Method Not Allowed
INFO:     127.0.0.1:13535 - "POST /password HTTP/1.1" 422 Unprocessable Entity
INFO:     127.0.0.1:13564 - "POST /password HTTP/1.1" 200 OK
INFO:     127.0.0.1:13564 - "POST /password HTTP/1.1" 400 Bad Request
INFO:     127.0.0.1:13589 - "POST /password HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218210410.png)

![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218210351.png)

In [45]:
from fastapi import FastAPI, Body, HTTPException

app = FastAPI()


@app.post("/password")
async def check_password(password: str = Body(...), password_confirm: str = Body(...)):
    if password != password_confirm:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            detail={
                "message": "Passwords don't match.",
                "hints": [
                    "Check the caps lock on your keyboard",
                    "Try to make the password visible by clicking on the eye icon to check your typing."
                ]
            },
        )
    return {
        "message": "Passwords match"
    }

In [46]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:13708 - "POST /password HTTP/1.1" 200 OK
INFO:     127.0.0.1:13708 - "POST /password HTTP/1.1" 400 Bad Request


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218210716.png)



## Building a custom response

Most of the time, you'll let FastAPI take care of building an HTTP response by simply
providing it with some data to serialize. Under the hood, FastAPI uses a subclass of
Response, called JSONResponse

However, there are other response classes that cover common cases:
- HTMLResponse -- return a html
- PlainTextResponse -- return raw text ( remembered the pyenv.run site?)
- RedirectResponse -- make a redirection
- StreamingResponse -- stream a flow of bytes
- FileResponse -- automatically build a proper file response given the path of a file on the disk

You have two ways of using them: either setting the
argument on the
response_class
path decorator or directly returning a response instance.

### using the response class argument

In [48]:
from fastapi import FastAPI
from fastapi.responses import HTMLResponse, PlainTextResponse

app = FastAPI()


@app.get("/html", response_class=HTMLResponse)
async def get_html():
    return """
<html>
    <head>
        <title>Hello world!</title>
    </head>
    <body>
        <h1>Hello world!</h1>
    </body>
</html>
    """


@app.get('/text', response_class=PlainTextResponse)
async def get_text():
    return """
#!/bin/bash
#
# Usage: curl https://pyenv.run | bash
#
# For more info, visit: https://github.com/pyenv/pyenv-installer
#
index_main() {
    set -e
    curl -s -S -L https://raw.githubusercontent.com/pyenv/pyenv-installer/master/bin/pyenv-installer | bash
}

index_main
    """

In [49]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:14000 - "POST /password HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:14007 - "POST /html HTTP/1.1" 405 Method Not Allowed
INFO:     127.0.0.1:14007 - "GET /html HTTP/1.1" 200 OK
INFO:     127.0.0.1:14055 - "GET /text HTTP/1.1" 200 OK
INFO:     127.0.0.1:14063 - "GET /text HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


Use return HTML method to return our model and images from the pyecharts?![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218211522.png)

Get a code content from the url and run it on bash directly ?
![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218211629.png)

### Making a redirection

In [54]:
from fastapi import FastAPI
from fastapi.responses import HTMLResponse, PlainTextResponse, RedirectResponse

app = FastAPI()


@app.get("/")
async def home():
    return {
        'helo': 'world!'
    }


@app.get("/redirect")
async def redirect():
    return RedirectResponse("/")

In [55]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:14241 - "GET /redirect HTTP/1.1" 307 Temporary Redirect
INFO:     127.0.0.1:14241 - "GET / HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


We got the results of '/' from '/redirect'

![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218212110.png)

In [56]:
from fastapi import FastAPI
from fastapi.responses import HTMLResponse, PlainTextResponse, RedirectResponse

app = FastAPI()


@app.get("/")
async def home():
    return {
        'helo': 'world!'
    }


@app.get("/redirect")
async def redirect():
    return RedirectResponse("/", status_code=status.HTTP_301_MOVED_PERMANENTLY)

In [57]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:14401 - "GET /redirect HTTP/1.1" 301 Moved Permanently
INFO:     127.0.0.1:14401 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:14401 - "GET /redirect HTTP/1.1" 301 Moved Permanently
INFO:     127.0.0.1:14401 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:14401 - "GET /redirect HTTP/1.1" 301 Moved Permanently
INFO:     127.0.0.1:14401 - "GET / HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


### saving  a file

In [59]:
!pip install aiofiles

In [86]:
from pathlib import Path
from fastapi import FastAPI
from fastapi.responses import FileResponse

app = FastAPI()


@app.get("/cat")
async def get_cat():
    # root_directory = path.dirname(path.dirname(__file__)) # father of current
    root_directory = Path.cwd()
    picture_path = root_directory.joinpath("assets", "cat.jpg")

    # return {
    #     'path': picture_path
    # }

    return FileResponse(picture_path)


In [87]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:4326 - "GET /cat HTTP/1.1" 200 OK
INFO:     127.0.0.1:4336 - "GET /cat HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218214735.png)

date: Sat, 18 Dec 2021 15:47:28 GMT
server: uvicorn
content-type: image/jpeg
content-length: 18184
last-modified: Sat, 18 Dec 2021 15:46:06 GMT
etag: 2c7ae8159617d23d57dad5fddbcdacb0

### custom response

In [88]:
from pathlib import Path
from fastapi import FastAPI
from fastapi.responses import FileResponse

app = FastAPI()


@app.get("/cat")
async def get_xml():
    content = """
<?xml version="1.0" encoding="UTF-8"?>
    <Hello>World</Hello>
    """

    return Response(content=content, media_type="application/xml")

In [89]:
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

INFO:     Started server process [12364]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:4483 - "GET /cat HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [12364]


date: Sat, 18 Dec 2021 15:50:47 GMT
server: uvicorn
content-length: 69
content-type: application/xml

## Structuring a bigger project with multiple routers

- logic
- data models
- API endpoints
- services
- ...

![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218215406.png)
![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211218215429.png)

At the root of the project, we have a file named
app.py
that will expose the main FastAPI application. Th
file defines a dummy database
db.py
for the sake of the example.

![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211219182254.png)

![](https://raw.githubusercontent.com/azataiot/images/master/2022/20211219182311.png)

Finally, the tags argument helps you to group endpoints in the interactive documentation for better readability.

__request__ :

- path
- query
- parameters
- headers
- body

**Response** :
- JSON response
- error
- file

